File: front/src/types/implementation-proposal.ts
export type ProposalStatus =
  | "draft"
  | "submitted"
  | "under_review"
  | "approved"
  | "rejected"
  | "revision_required";

export type ProposalPriority = "low" | "medium" | "high" | "critical";

export interface StatusHistory {
  id: number;
  from_status: ProposalStatus | null;
  to_status: ProposalStatus;
  comment: string | null;
  changed_at: string;
  changed_by: number;
}

export interface ImplementationProposal {
  id: number;
  dissertation_id: number | null;
  proposed_by: number;
  reviewed_by: number | null;
  title: string;
  problem_description: string;
  proposal_text: string;
  expected_result: string;
  implementation_area: string;
  implementation_org: string;
  priority: ProposalPriority;
  source_chapter: string | null;
  source_pages: string | null;
  status: ProposalStatus;
  submitted_at: string | null;
  reviewed_at: string | null;
  deadline: string | null;
  reviewer_comment: string | null;
  revision_notes: string | null;
  attachment_url: string | null;
  created_at: string;
  updated_at: string;
  status_history: StatusHistory[];
}

export interface ImplementationProposalList {
  items: ImplementationProposal[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export const STATUS_LABELS: Record<ProposalStatus, string> = {
  draft: "Qoralama",
  submitted: "Yuborildi",
  under_review: "Ko'rib chiqilmoqda",
  approved: "Tasdiqlandi",
  rejected: "Rad etildi",
  revision_required: "Qayta ishlash kerak",
};

export const STATUS_COLORS: Record<ProposalStatus, string> = {
  draft: "bg-gray-100 text-gray-700",
  submitted: "bg-blue-100 text-blue-700",
  under_review: "bg-yellow-100 text-yellow-700",
  approved: "bg-green-100 text-green-700",
  rejected: "bg-red-100 text-red-700",
  revision_required: "bg-orange-100 text-orange-700",
};

export const PRIORITY_LABELS: Record<ProposalPriority, string> = {
  low: "Past",
  medium: "O'rta",
  high: "Yuqori",
  critical: "Kritik",
};

export interface ProblemItem {
  id?: number;
  order_num: number;
  problem_text: string;
  source_page?: string;
}

export interface ProposalItem {
  id?: number;
  order_num: number;
  proposal_text: string;
  source_page?: string;
}

export interface SearchResult {
  id: string;
  title: string;
  author: string;
  university: string;
  year: number | null;
  degree: string;
  matched_problems: { text: string; page?: string }[];
  matched_proposals: { text: string; page?: string }[];
}

File: back/app/schemas/implementation_proposal.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.implementation_proposal import ProposalPriority, ProposalStatus


class ProposalStatusHistoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    from_status: ProposalStatus | None
    to_status: ProposalStatus
    comment: str | None
    changed_at: datetime
    changed_by: int


class ImplementationProposalCreate(BaseModel):
    """dissertation_id ixtiyoriy: avvalo muammo/yechim taklifi, tasdiqdan keyin dissertatsiya yaratiladi."""
    dissertation_id: int | None = None
    title: str = Field(..., min_length=5, max_length=500)
    problem_description: str = Field(..., min_length=20)
    proposal_text: str = Field(..., min_length=20)
    expected_result: str = Field(..., min_length=10)
    implementation_area: str
    implementation_org: str
    priority: ProposalPriority = ProposalPriority.medium
    source_chapter: str | None = None
    source_pages: str | None = None


class ImplementationProposalUpdate(BaseModel):
    dissertation_id: int | None = None
    title: str | None = None
    problem_description: str | None = None
    proposal_text: str | None = None
    expected_result: str | None = None
    implementation_area: str | None = None
    implementation_org: str | None = None
    priority: ProposalPriority | None = None
    source_chapter: str | None = None
    source_pages: str | None = None
    deadline: date | None = None
    internal_notes: str | None = None


class ReviewAction(BaseModel):
    comment: str | None = None


class RejectAction(BaseModel):
    comment: str = Field(..., min_length=10)


class RevisionAction(BaseModel):
    revision_notes: str = Field(..., min_length=10)


class ImplementationProposalOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    dissertation_id: int | None
    proposed_by: int
    reviewed_by: int | None
    title: str
    problem_description: str
    proposal_text: str
    expected_result: str
    implementation_area: str
    implementation_org: str
    priority: ProposalPriority
    source_chapter: str | None
    source_pages: str | None
    status: ProposalStatus
    submitted_at: datetime | None
    reviewed_at: datetime | None
    deadline: date | None
    reviewer_comment: str | None
    revision_notes: str | None
    attachment_url: str | None
    created_at: datetime
    updated_at: datetime
    status_history: list[ProposalStatusHistoryOut] = []


class ImplementationProposalList(BaseModel):
    items: list[ImplementationProposalOut]
    total: int
    page: int
    size: int
    pages: int
